debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerStreamingChangeEventSource.java

```diff
@@ -0,0 +1,263 @@
+
+                startScn = offsetContext.getScn();
+                createAuditTable(connection);
+                LOGGER.trace("current millis {}, db time {}", System.currentTimeMillis(), getTimeDifference(connection));
+                transactionalBufferMetrics.setTimeDifference(new AtomicLong(getTimeDifference(connection)));
+
+                if (!isContinuousMining && startScn < getFirstOnlineLogScn(connection)) {
+                    throw new RuntimeException("Online REDO LOG files don't contain the offset SCN. Clean offset and start over");
+                }
+
+                // 1. Configure Log Miner to mine online redo logs
+                setNlsSessionParameters(jdbcConnection);
+                checkSupplementalLogging(jdbcConnection, connectorConfig.getPdbName());
+
+                if (strategy == OracleConnectorConfig.LogMiningStrategy.CATALOG_IN_REDO) {
+                    buildDataDictionary(connection);
+                }
+
+                if (!isContinuousMining) {
+                    setRedoLogFilesForMining(connection, startScn);
+                }
+
+                LogMinerQueryResultProcessor processor = new LogMinerQueryResultProcessor(context, logMinerMetrics, transactionalBuffer,
+                        dmlParser, offsetContext, schema, dispatcher, transactionalBufferMetrics, catalogName, clock);
+
+                // 2. Querying LogMiner view while running
+                Set<String> currentRedoLogFiles = getCurrentRedoLogFiles(connection, logMinerMetrics);
+                while (context.isRunning()) {
+
+                    endScn = getEndScn(connection, startScn, logMinerMetrics);
+                    // LOGGER.trace("startScn: {}, endScn: {}", startScn, endScn);
+
+                    metronome = Metronome.sleeper(Duration.ofMillis(logMinerMetrics.getMillisecondToSleepBetweenMiningQuery()), clock);
+                    metronome.pause();
+
+                    Set<String> possibleNewCurrentLogFile = getCurrentRedoLogFiles(connection, logMinerMetrics);
+                    if (!currentRedoLogFiles.equals(possibleNewCurrentLogFile)) {
+                        LOGGER.debug("\n\n***** SWITCH occurred *****\n" + " from:{} , to:{} \n\n", currentRedoLogFiles, possibleNewCurrentLogFile);
+
+                        // This is the way to mitigate PGA leak.
+                        // With one mining session it grows and maybe there is another way to flush PGA, but at this point we use new mining session
+                        endMining(connection);
+
+                        if (!isContinuousMining) {
+                            if (strategy == OracleConnectorConfig.LogMiningStrategy.CATALOG_IN_REDO) {
+                                buildDataDictionary(connection);
+                            }
+
+                            abandonOldTransactionsIfExist(connection);
```

Comment:
I put the miners in a cluster and store abandoned transactions in Hazelcast for later retrieval. This is a bit of shame to throw the transaction away. And what happens if the transaction is huge? Then the switch keeps happening before the transaction is handled. Or did you handle that situation somewhere else?

Reply:
From what I can gather from the code, I suspect the transaction is then incomplete in Kafka.  @AndreyIg is that accurate?

Reply:
OLTP supposed to use short transactions. If a transaction lasts hours, there is something wrong in the application design in my opinion. Without abandoning mechanism, we might end up with full re-snapshot after restarting the connector.
Huge transaction should not be a problem,as long as it is fits let say in one hour